### ISSUE TITLE
Pupil plane sampling for WFIRST PSFs

### ISSUE THREAD

**Issue by arunkannawadi** [GalSim-developers/GalSim#772]
On trying to generate the PSF profiles for a given band, I get the following warning:

```
Input pupil plane image may not be sampled well enough!
Consider increasing sampling by a factor 98.131897, and/or check PhaseScreenPSF outputs for signs of folding in real space.
```
The sampling factor to be increased is ~100 ?! Here's a code snippet to get that warning:

```
import galsim
import galsim.wfirst as wfirst
PSFs = wfirst.getPSF(SCAs=7, wavelength=1840.)
```

**Comment by jmeyers314** [comment]
Hi Arun,

This is part of the new API for OpticalPSF, which has stricter requirements for how to set the scale (in meters) for an image of the pupil plane.  Sorry that I didn't think about how this might affect the wfirst module.  From the `galsim.Aperture` documentation:

```
The pixel scale of the pupil plane can be specified in one of three ways.  In descending order 
of priority, these are:
      1.  The `pupil_plane_scale` keyword argument (units are meters).
      2.  The `pupil_plane_im.scale` attribute (units are meters).
      3.  If (1) and (2) are both None, then the scale will be inferred by assuming that the
          illuminated pixel farthest from the image center is at a physical distance of self.diam/2.
```

In this case, the image file of the WFIRST focal plane has a scale of 1.0 (I'm not sure if this is from the WCS or if it's a default), so the 4096x4096 image is interpreted as having physical extent of ~4km x ~4km.  While it would certainly be cool to have a ~km aperture optical/infrared telescope in space, I don't think this is in the plans for WFIRST!

Of the three ways listed above for setting the correct scale, I think in this case setting it as an image attribute (via the image file header WCS) makes the most sense.  (Anyone know if it's possible to indicate in the FITS header that the WCS units are in meters?)

That said, it would probably be a good idea to check in the code if the inferred diameter coming from the pupil image + scale at least roughly matches the diameter specified as an argument, and report a helpful warning if they don't.

Let me know if you want to tackle this one, or if you'd like me to.  If you want me to though, I'll need to know the precise pixel scale for the WFIRST pupil plane image.

**Comment by rmjarvis** [comment]
Hm.  Seems like we should probably resolve this for version 1.4.

The WFirst mirror diameter is 2.4 m.  And the pupil image has non-zero values from x=1202 to x=2897. So the proper scale here is probably 2.4 m / 1696 = 0.00141509.

The simplest solution is probably to just provide this value as pupil_plane_scale when using this file.

**Comment by rmjarvis** [comment]
I made a stab at this, and I noticed that there are no tests of either `high_accuracy=True` or `approximate_struts=False` in the test_wfirst.py test suite, which is why this never got caught.  Such tests are slow, but we usually do those kinds of tests in `if __name__ == "__main__"` blocks.

So I added that, and the tests with the pupil plane are no longer giving the warning that Arun found.  But they are failing.  I'm not sure if this is pointing to a bug in the code (either the wfirst module or the new OpticalPSF stuff) or in my proposed test.  So hopefully @arunkannawadi and @jmeyers314 could take a look and see if what I did makes sense.

**Comment by rmjarvis** [comment]
For reference, the images that I thought should be equal, but which are not equal for `approximage_struts=False, high_accuracy=False' (It doesn't get to the last case with `high_accuracy=True`.) are the following:

<img width="714" alt="wfirst_psf" src="https://cloud.githubusercontent.com/assets/623887/16835670/09db14de-49ba-11e6-9a49-e95d012c663f.png">

**Comment by jmeyers314** [comment]
I was able to make your new tests pass by modifying the evaluateAtWavelength method of ChromaticOpticalPSF to forward the wavelength and diameter as separate kwargs instead of using lam_over_diam.  However, I don't yet understand why the latter didn't also work, so I'll keep investigating.

Also, looks like galsim.wfirst.diameter is 2.36 not 2.4?  That's part of why I was hoping Arun could tell us the precise scale of the pupil image.

**Comment by arunkannawadi** [comment]
Hi Josh,

The ChromaticOpticalPSF does not seem to use the new API with
`galsim.Aperture`, so I wouldn't expect to see this warning while
evaluating it at a particular wavelength. I haven't had a look at the new
code, but I can keep looking into it.

The relevant WFIRST parameters should be available in __init__ file within
the `wfirst` module. So the diameter is 2.36m and the pixel scale is 0.11
arcsec/pixel.

On Thu, Jul 14, 2016 at 12:02 PM, Josh Meyers <notifications@github.com>
wrote:

> I was able to make your new tests pass by modifying the
> evaluateAtWavelength method of ChromaticOpticalPSF to forward the
> wavelength and diameter as separate kwargs instead of using lam_over_diam.
> However, I don't yet understand why the latter didn't also work, so I'll
> keep investigating.
>
> Also, looks like galsim.wfirst.diameter is 2.36 not 2.4? That's part of
> why I was hoping Arun could tell us the precise scale of the pupil image.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GalSim-developers/GalSim/issues/772#issuecomment-232710692>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE3fHn9CFo5o_TOg3oPTbuXwkVixKXciks5qVl2rgaJpZM4JLuPS>
> .
>



-- 
Regards

Arun Kannawadi Jayaraman
PhD candidate, Dept. of Physics,
Carnegie Mellon University,
Pittsburgh, PA 15213

**Comment by jmeyers314** [comment]
Hi @arunkannawadi, 

Part of the new API included replacing `OpticalPSF`, which `ChromaticOpticalPSF` does use.  So that's where the `Aperture` stuff gets pulled into this (via some intermediate layers).

Also, 0.11 arcsec/pixel must be the scale of the focal plane, not the pupil plane.  What the new API requires is to know how large the pixels are in meters for the file WFIRST-AFTA_Pupil_Mask_C5_20141010_PLT.fits.gz.  @rmjarvis's estimate above is probably pretty close (maybe substituting 2.36 for 2.4), but I wasn't sure if you had a more precise number or not.

**Comment by rmandelb** [comment]
> Also, 0.11 arcsec/pixel must be the scale of the focal plane, not the pupil plane.

Yes.

@rmjarvis 's estimate is pretty good (with the 2.4->2.36) but I just asked Jeff Kruk, who provided the images, if it would be possible for the pixel scale to be put in the header in future versions - and if he can tell it to us now for use with these images.

**Comment by rmandelb** [comment]
> Seems like we should probably resolve this for version 1.4.

+1

**Comment by rmandelb** [comment]
I think for now we should go with Mike's approximate estimate.  The WFIRST project only provided pupil plane images in .tif format, and I had to both edit them (they had spurious nonzero pixels outside the pupil) and convert them to FITS.  in practice, while a pixel scale was given for the .tif images, the FITS ones have a different pixel scale, so my conversion makes the information given with the .tif files useless.

Jeff says that in future iterations, they will provide some FITS files and will make sure that the keywords are informative as to the pixel scale.  But for now I think we should set it ourselves using Mike's suggested calculation.

**Comment by rmjarvis** [comment]
> I was able to make your new tests pass by modifying the evaluateAtWavelength method of ChromaticOpticalPSF to forward the wavelength and diameter as separate kwargs instead of using lam_over_diam. However, I don't yet understand why the latter didn't also work, so I'll keep investigating.

I looked into this a bit this afternoon, and I think the change we want is to disallow providing `lam_over_diam` when also providing a pupil plane image with a definite scale.  

I'm not positive whether it is ok if the scale is determined automatically, but if the scale is provided, then I think [this line](https://github.com/GalSim-developers/GalSim/blob/master/galsim/phase_psf.py#L1035) is invalid when the `lam` value is taken to be the arbitrary `lam=500` that happens when `lam_over_diam` is given.

If the pupil plane is constructed from struts and obscuration, etc. then this is fine, since they are constructed in the context of this arbitrary wavelength.  I think it's probably also fine if the pupil plane image doesn't have a set scale and it's just derived from the image and `diam` (which is also abitrary as `500 / lam_over_diam`).  But if the user provides a particular scale to the pupil plane image, then we can't get away with using an arbitrary `diam`.

**Comment by rmjarvis** [comment]
Closed via PR #778
